src/main/java/com/mraof/minestuck/alchemy/GristGutter.java

```diff
@@ -0,0 +1,116 @@
+package com.mraof.minestuck.alchemy;
+
+import com.mraof.minestuck.skaianet.Session;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.level.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import software.bernie.shadowed.eliotlash.mclib.math.functions.limit.Min;
+
+import java.util.Map;
+
+/**
+ * A class that handles Grist overflow whenever you aqcuire too much grist.
+ * @author Doro
+ */
+public class GristGutter extends GristSet
+{
+	private static Session session;
+	public int gutterTotal = -1;
+	public static final int GUTTER_CAPACITY = 10000;
+	private static final Logger LOGGER = LogManager.getLogger();
+	public void setSession(Session session)
+	{
+		this.session = session;
+	}
+	
+	public double getGutterCapacity(Session session)
+	{
+		double gutMul = session.getGutterMultiplier();
+		double gutcap = (GUTTER_CAPACITY * gutMul);
+		return gutcap;
+	}
+	public long getGutterTotal()
+	{
+		if(gutterTotal < 0)
+		{
+			gutterTotal = 0;
+			for(Map.Entry<GristType, Long> pair : gristTypes.entrySet())
+			{
+				gutterTotal += pair.getValue();
+			}
+		}
+		return gutterTotal;
+	}
+	public GristSet gristToSpill = new GristSet();
+	public void spillGrist(Level level, Player player)
+	{
+	//isn't being used by anything, but when i delete it everything breaks
+		gristToSpill.spawnGristEntities(level, player.getX(), player.getY(), player.getZ(), level.random, entity -> entity.setDeltaMovement(entity.getDeltaMovement().multiply(1.5, 0.5, 1.5)), 90, level.random.nextInt(6) > 0 ? 1 : 2);
+	}
+	@Override
+	public GristGutter addGrist(GristType type, long amount)
+	{
+		if(type != null)
+		{
+			GristSet sOverflowGrist = new GristSet();//creates a new gristset called Super overflow
+			long originalAmount = this.gristTypes.getOrDefault(type, 0L);
+			long maximumAllowed = (long) (gutterTotal - getGutterCapacity(session) + originalAmount);
+			
+			this.gristTypes.compute(type, (key, value) -> value == null ? amount : value + amount);
+			gutterTotal += amount;//adds grist to gutter
+			
+			//logger
+			LOGGER.debug("Gutter after adding " + amount + " " + type.getDisplayName().toString() + " grist:");
+			LOGGER.debug("Total: " + getGutterTotal());
+			for(GristType t : this.gristTypes.keySet())
+			{
+				LOGGER.debug(t.getDisplayName().toString() + ": " + gristTypes.get(t));
```

Comment:
Theres a lot of errors all over GristGutter pertaining to "gristTypes" a piece of data in GristSet. Instead of GristGutter extending GristSet, perhaps it should contain GristSet as one of its parameters

Reply:
no clue on how to do that

Reply:
Agreed that the `GristSet` shouldn't be extended for this class. The easiest way to go from here is to have a `GristSet` field which `GristGutter` uses operations on.
Direct uses of `this.gristTypes` would instead use various different functions on the grist set, which should be straightforward. The approach to `addGrist()` also need to change, which is slightly less straightforward, but still pretty doable. Currently, `gristGutter.addGrist(GristSet)` is used in two places, with `gristGutter.addGrist(GristType, long)` being the one that is overridden. If the former was added to the `GristGutter` class to call the latter, that should solve the problem straightforwardly.